.azure-pipelines/ultimate-pipeline.yml

```diff
@@ -898,6 +945,140 @@ stages:
+          fi
+
+          add_to_matrix () {
+            name="${baseImage//./_}_${publishTargetFramework//./_}_${explorationTestUseCase//./_}_${explorationTestName}"
+            matrix="${matrix} '${name}':"
+            matrix="${matrix} { 'baseImage': '${baseImage}', "
+            matrix="${matrix} 'publishTargetFramework': '${publishTargetFramework}', "
+            matrix="${matrix} 'explorationTestUseCase': '${explorationTestUseCase}', "
+            matrix="${matrix} 'explorationTestName': '${explorationTestName}' },"
+          }
+
+          for baseImage in "debian" "alpine" ; do
+            for explorationTestUseCase in ${explorationTestUseCases[@]} ; do
+
+              publishTargetFramework="netcoreapp2.1"
+              for explorationTestName in "eshoponweb" "protobuf" ; do
+                add_to_matrix
+              done
+
+              publishTargetFramework="netcoreapp3.1"
+              explorationTestName="cake"
+              add_to_matrix
+
+              publishTargetFramework="net5.0"
+              explorationTestName="cake"
+              add_to_matrix
+
+              publishTargetFramework="net6.0"
+              for explorationTestName in  "cake" "swashbuckle"; do
+                add_to_matrix
+              done
+            done
+          done
+          matrix="${matrix::-1} }"
+          echo "##vso[task.setVariable variable=values;isOutput=true]${matrix}"
+        name: mtrx
+
+
+  - job: Test
+    dependsOn: generator
+    strategy:
+      matrix: $[ dependencies.generator.outputs['mtrx.values'] ]
+
+    pool:
+      vmImage: ubuntu-18.04
+
+    steps:
+    # Doing a clean of obj files _before_ restore to remove build output from previous runs
+    # Can't do a full clean, as otherwise restore-working-directory fails
+    # Only necessary for ARM64, but shouldn't cause any harm on others
+    # Can't ifdef it as depends on a matrix variable
+    - template: steps/run-in-docker.yml
+      parameters:
+        build: true
+        baseImage: $(baseImage)
+        command: "CleanObjFiles"
+
+    - template: steps/restore-working-directory.yml
+      parameters:
+        artifact: build-linux-$(baseImage)-working-directory
+
+    - template: steps/run-in-docker.yml
+      parameters:
+        build: true
+        baseImage: $(baseImage)
+        command: "SetupExplorationTests -ExplorationTestUseCase $(explorationTestUseCase) --explorationTestName $(explorationTestName) --framework $(publishTargetFramework)"
+
+    - task: DockerCompose@0
+      displayName: docker-compose build ExplorationTests
+      env:
+        baseImage: $(baseImage)
+      inputs:
+        containerregistrytype: Container Registry
+        dockerComposeCommand: build --build-arg baseImage=$(baseImage) --build-arg explorationTestUseCase=$(explorationTestUseCase) --build-arg explorationTestName=$(explorationTestName) --build-arg framework=$(publishTargetFramework) ExplorationTests
```

Comment:
I recently moved away from using the built-in docker-compose command so that I could add retries to the commands (they often fail to pull for example) but I think we can cross that bridge if we come to it!